File: iphone app/Novios/Services/GameService.swift
import Foundation
import FirebaseFirestore

public class GameService: ObservableObject {
    public static let shared = GameService()

    private let db = Firestore.firestore()
    private let defaults = UserDefaults.standard

    private var coupleId: String {
        [CoupleService.diegoUid, CoupleService.yosmariUid].sorted().joined(separator: "_")
    }

    private var juegosRef: CollectionReference {
        db.collection("parejas").document(coupleId).collection("juegos")
    }

    private func typeRef(_ gameType: String) -> CollectionReference {
        juegosRef.document(gameType).collection("items")
    }

    // MARK: - Quizzes
    public func streamQuizzes() -> Query {
        typeRef("quizzes").order(by: "createdAt", descending: true)
    }

    public func saveQuiz(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("quizzes").document($0) } ?? typeRef("quizzes").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["authorId"] = defaults.string(forKey: "user_id") ?? ""
        d["createdAt"] = FieldValue.serverTimestamp()
        d["updatedAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteQuiz(_ id: String) {
        typeRef("quizzes").document(id).delete()
    }

    // MARK: - Truth or Dare
    public func streamTD(category: String) -> Query {
        typeRef("verdad_reto").whereField("category", isEqualTo: category)
    }

    public func saveTD(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("verdad_reto").document($0) } ?? typeRef("verdad_reto").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteTD(_ id: String) {
        typeRef("verdad_reto").document(id).delete()
    }

    // MARK: - Never Have I Ever
    public func streamNever(category: String? = nil) -> Query {
        var q: Query = typeRef("yo_nunca")
        if let cat = category { q = q.whereField("category", isEqualTo: cat) }
        return q
    }

    public func saveNever(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("yo_nunca").document($0) } ?? typeRef("yo_nunca").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteNever(_ id: String) {
        typeRef("yo_nunca").document(id).delete()
    }

    // MARK: - Would You Rather
    public func streamPrefer() -> Query {
        typeRef("que_prefieres")
    }

    public func savePrefer(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("que_prefieres").document($0) } ?? typeRef("que_prefieres").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deletePrefer(_ id: String) {
        typeRef("que_prefieres").document(id).delete()
    }

    // MARK: - Roulettes
    public func streamRoulettes() -> Query {
        typeRef("ruletas")
    }

    public func saveRoulette(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("ruletas").document($0) } ?? typeRef("ruletas").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteRoulette(_ id: String) {
        typeRef("ruletas").document(id).delete()
    }

    // MARK: - Hangman
    public func streamHangman() -> Query {
        typeRef("ahorcados")
    }

    public func saveHangmanWord(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("ahorcados").document($0) } ?? typeRef("ahorcados").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteHangmanWord(_ id: String) {
        typeRef("ahorcados").document(id).delete()
    }

    // MARK: - Dice
    public func streamDice() -> Query {
        typeRef("dados")
    }

    public func saveDice(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("dados").document($0) } ?? typeRef("dados").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteDice(_ id: String) {
        typeRef("dados").document(id).delete()
    }

    // MARK: - Love / Compatibility
    public func streamLoveQuestions() -> Query {
        typeRef("amor")
    }

    public func saveLoveQuestion(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef("amor").document($0) } ?? typeRef("amor").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    // MARK: - Stats
    public func saveGameStats(_ gameType: String, _ stats: [String: Any]) {
        var d = stats
        d["playerId"] = defaults.string(forKey: "user_id") ?? ""
        d["playerName"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["timestamp"] = FieldValue.serverTimestamp()
        juegosRef.document(gameType).collection("stats").addDocument(data: d)
    }

    public func streamGameStats(_ gameType: String) -> Query {
        juegosRef.document(gameType).collection("stats")
            .order(by: "timestamp", descending: true)
            .limit(to: 50)
    }

    // MARK: - Favorite / Duplicate
    public func toggleFavorite(gameType: String, itemId: String) {
        let ref = typeRef(gameType).document(itemId)
        ref.getDocument { snapshot, _ in
            guard let data = snapshot?.data() else { return }
            var favs = data["favoritedBy"] as? [String] ?? []
            let uid = self.defaults.string(forKey: "user_id") ?? ""
            if let idx = favs.firstIndex(of: uid) {
                favs.remove(at: idx)
            } else {
                favs.append(uid)
            }
            ref.updateData(["favoritedBy": favs])
        }
    }

    public func duplicateItem(gameType: String, itemId: String) {
        let ref = typeRef(gameType).document(itemId)
        ref.getDocument { snapshot, _ in
            guard var data = snapshot?.data() else { return }
            data.removeValue(forKey: "createdAt")
            data.removeValue(forKey: "updatedAt")
            data.removeValue(forKey: "favoritedBy")
            data["author"] = self.defaults.string(forKey: "user_name") ?? "Yo"
            data["authorId"] = self.defaults.string(forKey: "user_id") ?? ""
            data["createdAt"] = FieldValue.serverTimestamp()
            data["duplicatedFrom"] = itemId
            self.typeRef(gameType).addDocument(data: data)
        }
    }

    // MARK: - Collections
    public func streamCollections() -> Query {
        juegosRef.document("colecciones").collection("items")
    }

    public func saveCollection(_ data: [String: Any], id: String? = nil) {
        let ref = id.map { juegosRef.document("colecciones").collection("items").document($0) }
            ?? juegosRef.document("colecciones").collection("items").document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteCollection(_ id: String) {
        juegosRef.document("colecciones").collection("items").document(id).delete()
    }

    // MARK: - Generic
    public func streamAll(gameType: String) -> Query {
        typeRef(gameType)
    }

    public func saveItem(gameType: String, data: [String: Any], id: String? = nil) {
        let ref = id.map { typeRef(gameType).document($0) } ?? typeRef(gameType).document()
        var d = data
        d["author"] = defaults.string(forKey: "user_name") ?? "Yo"
        d["authorId"] = defaults.string(forKey: "user_id") ?? ""
        d["createdAt"] = FieldValue.serverTimestamp()
        ref.setData(d, merge: true)
    }

    public func deleteItem(gameType: String, id: String) {
        typeRef(gameType).document(id).delete()
    }
}

File: lib/services/game_service.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'local_storage.dart';
import 'couple_service.dart';

class GameService {
  static final GameService _instance = GameService._();
  factory GameService() => _instance;
  GameService._();

  final _db = FirebaseFirestore.instance;
  String get _coupleId => CoupleService.parejaId;

  // ─── Paths ───
  CollectionReference get _gamesRef =>
      _db.collection('parejas').doc(_coupleId).collection('juegos');

  CollectionReference _typeRef(String gameType) =>
      _gamesRef.doc(gameType).collection('items');

  // ─── Quizzes ───
  Stream<QuerySnapshot> streamQuizzes() =>
      _typeRef('quizzes').orderBy('createdAt', descending: true).snapshots();

  Future<void> saveQuiz(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('quizzes').doc(id) : _typeRef('quizzes').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'authorId': LocalStorage().getUserId() ?? '',
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteQuiz(String id) =>
      _typeRef('quizzes').doc(id).delete();

  // ─── Truth or Dare ───
  Stream<QuerySnapshot> streamTD(String category) =>
      _typeRef('verdad_reto').where('category', isEqualTo: category).snapshots();

  Future<void> saveTD(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('verdad_reto').doc(id) : _typeRef('verdad_reto').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteTD(String id) =>
      _typeRef('verdad_reto').doc(id).delete();

  // ─── Never Have I Ever ───
  Stream<QuerySnapshot> streamNever({String? category}) {
    Query q = _typeRef('yo_nunca');
    if (category != null) q = q.where('category', isEqualTo: category);
    return q.snapshots();
  }

  Future<void> saveNever(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('yo_nunca').doc(id) : _typeRef('yo_nunca').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteNever(String id) =>
      _typeRef('yo_nunca').doc(id).delete();

  // ─── Would You Rather ───
  Stream<QuerySnapshot> streamPrefer() =>
      _typeRef('que_prefieres').snapshots();

  Future<void> savePrefer(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('que_prefieres').doc(id) : _typeRef('que_prefieres').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deletePrefer(String id) =>
      _typeRef('que_prefieres').doc(id).delete();

  // ─── Roulettes ───
  Stream<QuerySnapshot> streamRoulettes() =>
      _typeRef('ruletas').snapshots();

  Future<void> saveRoulette(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('ruletas').doc(id) : _typeRef('ruletas').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteRoulette(String id) =>
      _typeRef('ruletas').doc(id).delete();

  // ─── Hangman ───
  Stream<QuerySnapshot> streamHangman() =>
      _typeRef('ahorcados').snapshots();

  Future<void> saveHangmanWord(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('ahorcados').doc(id) : _typeRef('ahorcados').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteHangmanWord(String id) =>
      _typeRef('ahorcados').doc(id).delete();

  // ─── Custom Dice ───
  Stream<QuerySnapshot> streamDice() =>
      _typeRef('dados').snapshots();

  Future<void> saveDice(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('dados').doc(id) : _typeRef('dados').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> deleteDice(String id) =>
      _typeRef('dados').doc(id).delete();

  // ─── Love / Compatibility ───
  Stream<QuerySnapshot> streamLoveQuestions() =>
      _typeRef('amor').snapshots();

  Future<void> saveLoveQuestion(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef('amor').doc(id) : _typeRef('amor').doc();
    await ref.set({
      ...data,
      'author': LocalStorage().getUserName() ?? 'Yo',
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  // ─── Stats ───
  Future<void> saveGameStats(String gameType, Map<String, dynamic> stats) async {
    await _gamesRef.doc(gameType).collection('stats').add({
      ...stats,
      'playerId': LocalStorage().getUserId() ?? '',
      'playerName': LocalStorage().getUserName() ?? 'Yo',
      'timestamp': FieldValue.serverTimestamp(),
    });
  }

  Stream<QuerySnapshot> streamGameStats(String gameType) =>
      _gamesRef.doc(gameType).collection('stats')
          .orderBy('timestamp', descending: true)
          .limit(50)
          .snapshots();

  // ─── Favorite / Duplicate ───
  Future<void> toggleFavorite(String gameType, String itemId) async {
    final doc = _typeRef(gameType).doc(itemId);
    final snap = await doc.get();
    if (!snap.exists) return;
    final data = snap.data() as Map<String, dynamic>;
    final currentFavs = List<String>.from(data['favoritedBy'] ?? []);
    final uid = LocalStorage().getUserId() ?? '';
    if (currentFavs.contains(uid)) {
      currentFavs.remove(uid);
    } else {
      currentFavs.add(uid);
    }
    await doc.update({'favoritedBy': currentFavs});
  }

  Future<void> duplicateItem(String gameType, String itemId) async {
    final doc = _typeRef(gameType).doc(itemId);
    final snap = await doc.get();
    if (!snap.exists) return;
    final data = snap.data() as Map<String, dynamic>;
    final newData = Map<String, dynamic>.from(data)
      ..remove('createdAt')
      ..remove('updatedAt')
      ..remove('favoritedBy');
    newData['author'] = LocalStorage().getUserName() ?? 'Yo';
    newData['authorId'] = LocalStorage().getUserId() ?? '';
    newData['createdAt'] = FieldValue.serverTimestamp();
    newData['duplicatedFrom'] = itemId;
    await _typeRef(gameType).add(newData);
  }

  // ─── Generic stream all items for a game type ───
  Stream<QuerySnapshot> streamAll(String gameType) =>
      _typeRef(gameType).snapshots();

  // ─── Generic CRUD ───
  Future<void> saveItem(String gameType, Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _typeRef(gameType).doc(id) : _typeRef(gameType).doc();
    data['author'] = LocalStorage().getUserName() ?? 'Yo';
    data['authorId'] = LocalStorage().getUserId() ?? '';
    data['createdAt'] = FieldValue.serverTimestamp();
    await ref.set(data, SetOptions(merge: true));
  }

  Future<void> deleteItem(String gameType, String id) =>
      _typeRef(gameType).doc(id).delete();

  // ─── Collections / Folders ───
  CollectionReference get _collectionsRef =>
      _gamesRef.doc('colecciones').collection('items');

  Stream<QuerySnapshot> streamCollections() =>
      _collectionsRef.snapshots();

  Future<void> saveCollection(Map<String, dynamic> data, {String? id}) async {
    final ref = id != null ? _collectionsRef.doc(id) : _collectionsRef.doc();
    data['author'] = LocalStorage().getUserName() ?? 'Yo';
    data['authorId'] = LocalStorage().getUserId() ?? '';
    data['createdAt'] = FieldValue.serverTimestamp();
    await ref.set(data, SetOptions(merge: true));
  }

  Future<void> deleteCollection(String id) =>
      _collectionsRef.doc(id).delete();
}
